added more information on gc and rooting (#212)
servo/book: src/design-documentation/script.md

# Script

Servo is unique in that it uses garbage collection for some things that are non obvious. For example, every dom object

> (a struct with `#[dom_object])` on it is controlled by spidermonkeys garbage collector. Extra care is to be taken when interacting with this. While the garbage collector is complicated and has multiple modes we will assume for now the following.

Reviewer: dom_struct

> (a struct with `#[dom_object])` on it is controlled by spidermonkeys garbage collector. Extra care is to be taken when interacting with this. While the garbage collector is complicated and has multiple modes we will assume for now the following.

Reviewer: what extra care?

Author: added that this is what the document is about.

Whenever the GC runs, the rust program is stopped at whatever state it was in.

## Rooting and you

The most often used method is Rooting. Rooting (`Root<Dom<T>>`) means that we tell the garbage collector two things.
- Please do not delete any value that is reachable by this root.
- Please fix any pointer that points to this after you do a garbage collection.

In the following we will look at some code that is close to what is happening behind some abstractions but not the complete picture.
Let us now assume we have the following code:
```rust
#[dom_object]
struct Kittens {
    children: Vec<Cat>
}

fn some_function(cats: &Kittens) {
    let children = & cats.children;
    play_with(children)
}
```

Without rooting the GC could pause our program after we got the reference to children, move the `Kittens` struct around
and we would have invalid access when playing with them! This holds even if Kittens was rooted by some other mechanism earlier in the call chain.
Hence, we need to transform the code into
```rust

> #[dom_object]

Reviewer: dom_struct no?

struct Kittens {

>     children: Vec<Cat>

Reviewer: I think it would be a stronger example if Cat was also a dom object, and you would use not Vec but Vec<Dom<Cat>>, or even `RefCell<Vec<Dom<Cat>>>`.

Author: I agree making it a Vec<Dom<Cat>>.
I don't know about RefCell though. There is a whole other section about borrow checking and RefCell which explains this very well.. I am going to link to that.

}

fn some_function(cats: &Kittens) {

>     let children = Root::from_ref(&cats.children);

Reviewer: Here you could do let children: DomRoot<Cat> = cats.children.iter().map(|cat| cat.as_rooted()).collect();

    play_with(children)
}
```
Unrooting (the reverse) will automatically be done via a Drop impl on Root.

When you implement some HTML api you will generally only encounter rooted things (such as 'DomRoot\<Node\>'). And rooting

> and unrooting is generally pretty fast but there is a catch.

Reviewer: what catch?

Author: Removed this.

## CanGC and Crown
Now rooting could be easily forgotten when implementing particular exciting apis. How to we prevent this?
Crown is the answer (if you are working on Script things you should run `./mach build --use-crown` to be sure it is checking things.
In essence, Crown just checks that you do not forgotten to root things and you will sometimes see certain lints in the code talking to crown such as `#[cfg_attr(crown, allow(crown::unrooted_must_root))]`. These essentially disable crown and should only be used in very specific circumstances.

But there is another piece of the puzzle. CanGC.

> Some methods interacting with the DOM will never have any change of starting GC operations. So we can relax our requirement there.

Reviewer: any chance

> As the name implies, if a function takes this argument it means that it can have a GC operation in the middle of it and crown needs to check it.

Reviewer: Is that how it works? I thought CanGC was mostly about avoiding a mutable borrow, since during a GC Spidermonkey will trace all `Dom`, and in the case of `RefCell<Vec<Dom<T>>>`, then the tracer will take an immutable borrow which would panic if there is an outstanding mutable borrow.

Author: Is that the reason? I kind of assumed that crown wouldn't require rooting for things that have no CanGC.
If you are sure then I can trust you and reword that section to link to the borrow hazarding article.

Reviewer 2: Yes, CanGC is not used by crown, since it's only a best-effort annotation.

Reviewer: @jdm

```rust
fn some_function(cats: &Kittens, can_gc: CanGc) {
    let children = Root::from_ref(cats.children);
    play_with_cats(children);
    cleanup_after_cats(children, can_gc);
}
```

The pecise notion of this is that we can be sure that no GC will happen when we call 'play_with_cats' but there

> might be a GC happening while we call 'cleanup_after_cats'. The reason why some methods can incur GC and some methods do not are deep in the SpiderMonkey connections to servo and you will generally be obvious.

Reviewer: and generally not be obvious.

> In essence, you can not use can_gc until you call a method that needs it and then you go back along your callstack and give it as argument to every call until you arrive in something origininating from 'Bindings.conf' (see how to implement a dom meth'od). 

Reviewer: method.

You will sometimes see '&JSContext' and &mut JSContext' and 'NoGC'. You can think of '&mut JSContext' as a 'CanGc' and 'JSContext' and 'NoGC' as the absence of 'CanGc'.

> JSContext and associates are the new way as they have some advantages.

Reviewer: what advantages?

# The performance of rooting

```rust
#[dom_object]
struct Kittens {
    children: Vec<Cat>
}

fn some_function(cats: &Kittens) {
    let happy = cats.children.iter().map(|cat| Root::from_ref(cat)).any(|cat| cat.is_purring());
    if happy {
        println!("Happy cat found!");
    }
}
```

Now if the litter is especially large, we do a lot of work just rooting and unrooting the cats only to test a probably fast bool. Can we do something about this?

Yes. Remember rusts golden rule that &mut can only be hold exactly once!
What if we have a new type
```rust
struct UnrootedDom<'a> {
    inner_cat: Cat,
    js_context: &mut 'a JSContext,
}
```

> Then whenever we have a handle on `UnrootedDom` we know that as long as this lifes we can not call any methods that has a GC.

Reviewer: lives

The following is **invalid** code.
```rust
#[dom_object]
struct Kittens {
    children: Vec<UnrootedDom<'_>>,
}

fn make_cats(cx: &mut JSContext) -> Kittens {}

> fn play_with_cat(cx: &JSContext, cat: Cat) {}

Reviewer: &Cat

fn cleanup_after_cats(cx: &mut JSContext, cat: Cat) {}

fn some_function(cx: &mut JSContext) {
    let cats = make_cats(cx);
    for cat in cats.children {
        play_with_cat(cx, cat.inner_cat)
    }
    for cat in cats.children {
        cleanup_after_cats(cx, cat)
    }
}
```

> We will get a compile error that cx is borrowed mutabil once by 'make_cats' and once by 'cleanup_after_cats'.

Reviewer: mutably.

But notice that the call to 'play_with_cat' is perfectly fine.

This is the essence of 'UnrootedDomNode' and similar 'Unrooted' methods. Some of these might also take the 'NoGC' argument
```rust
fn play_with_cats<'a>(no_gc: &'a NoGC, cat: Cat) {}

> fn some_function(cx: &mut JSContext, cat: UnrootedDom<'_>) {

Reviewer: I think in general you don't want to pass the wrapper, just a & to the value wrapped, so &Cat.

# Script

Servo is unique in that it uses garbage collection for some things that are non obvious. For example, every dom object
(a struct with `#[dom_struct])` on it is controlled by spidermonkeys garbage collector. Extra care is to be taken when interacting with this. Which is what this document is about. While the garbage collector is complicated and has multiple modes we will assume for now the following.
Whenever the GC runs, the rust program is stopped at whatever state it was in.

> We will for now only assume read only access to objects as this makes it easer to understand.

Reviewer 2: ```suggestion
Servo is unique in that it uses garbage collection for some things that are non-obvious.
For example, every DOM object (a struct with `#[dom_struct])` is owned by SpiderMonkey's garbage collector.
This necessitates that the engine APIs that interact with these objects must be sound when garbage collection can occur at many points in the program.
While [the garbage collector](https://firefox-source-docs.mozilla.org/js/gc.html) is complex and has multiple modes, we can assume that whenever the GC runs then the Rust program does not run concurrently.
```

> ## Rooting and you

Reviewer 2: ```suggestion
## Rooting and rooted types
```

The most often used method is Rooting. Rooting (`Root<Dom<T>>`) means that we tell the garbage collector two things.
- Please do not delete any value that is reachable by this root.

> - Please fix any pointer that points to this after you do a garbage collection.

Reviewer 2: ```suggestion
An important part of garbage collection is establishing the roots of the object graph.
A root tells the garbage collector two things:
1. Do not delete any value that is transitively reachable from this root.
2. If garbage collection causes this rooted value to be moved in memory, all pointers to this value should be updated.

In Servo's code, roots are created automatically by using the `DomRoot<T>`, `Root<T>`, and `Rooted<T>` types.
See the [module documentation](https://doc.servo.org/script/dom/bindings/root/index.html) for more details.
```

> In the following we will look at some code that is close to what is happening behind some abstractions but not the complete picture.

Reviewer 2: ```suggestion
The following examples contain simplifications of real Servo code patterns to make them easier to understand.
```

> Let us now assume we have the following code:

Reviewer 2: ```suggestion
Assume the following code, declaring a `Kittens` type that contains pointers to `Cat` types, all of which are owned by the garbage collector:
```

```rust
#[dom_struct]
struct Kittens {
    children: Vec<Dom<Cat>>
}

> fn some_function(cats: &Kittens) {

Reviewer 2: ```suggestion
fn play_with_kittens(kittens: &Kittens) {
```

    let children = & cats.children;
    play_with(children)
}
```

Without rooting the GC could pause our program after we got the reference to children, move the `Kittens` struct around
and we would have invalid access when playing with them! This holds even if Kittens was rooted by some other mechanism earlier in the call chain.
Hence, we need to transform the code into
```rust
#[dom_struct]
struct Kittens {
    children: Vec<Dom<Cat>>
}

> fn some_function(cats: &Kittens) {

Reviewer 2: ```suggestion
fn play_with_kittens(cats: &Kittens) {
```

>     let children: DomRoot<Vec<Dom<Cat>>> = cats.children().iter().map(|cat| cat.as_rooted()).collect();

Reviewer 2: ```suggestion
    let children: Vec<DomRoot<Cat>> = cats.children.iter().map(|cat| cat.as_rooted()).collect();
```

    play_with(children)
}
```

> Unrooting (the reverse) will automatically be done via a Drop impl on Root.

Reviewer 2: ```suggestion
Unrooting (the reverse) will automatically be done via a Drop impl on `DomRoot<T>`.
```

When you implement some HTML api you will generally only encounter rooted things (such as 'DomRoot\<Node\>'). And rooting

> and unrooting is generally pretty fast.

Reviewer 2: ```suggestion
To ensure web API implementations in Servo are sound, the engine code is conservative and usually returns rooted types (such as `DomRoot<Node>`).
This is an acceptable tradeoff between safety and performance because rooting and unrooting operations are efficient.
```

    play_with_cats(children);
    cleanup_after_cats(children, can_gc);
}
```

The pecise notion of this is that we can be sure that no GC will happen when we call 'play_with_cats' but there
might be a GC happening while we call 'cleanup_after_cats'. The reason why some methods can incur GC and some methods do not are deep in the SpiderMonkey connections to servo and you will generally not be obvious.
In essence, you can not use can_gc until you call a method that needs it and then you go back along your callstack and give it as argument to every call until you arrive in something origininating from 'Bindings.conf' (see how to implement a dom meth'od).

You will sometimes see '&JSContext' and &mut JSContext' and 'NoGC'. You can think of '&mut JSContext' as a 'CanGc' and 'JSContext' and 'NoGC' as the absence of 'CanGc'.

JSContext and associates are the new way as they have some advantages.

# The performance of rooting

```rust
#[dom_struct]
struct Kittens {
    children: Vec<Cat>
}

fn some_function(cats: &Kittens) {
    let happy = cats.children.iter().map(|cat| Root::from_ref(cat)).any(|cat| cat.is_purring());
    if happy {
        println!("Happy cat found!");
    }
}
```

> Now if the litter is especially large, we do a lot of work just rooting and unrooting the cats only to test a probably fast bool. Can we do something about this?

Reviewer 2: ```suggestion
This API for interacting with kittens is safely rooted, but when there are many children the rooting and unrooting work can add up.
```
